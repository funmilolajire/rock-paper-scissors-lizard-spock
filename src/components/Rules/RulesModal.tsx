import ReactDOM from 'react-dom';
import { ModalContent } from './Rules--ModalContent';

export const RulesModal = () => {
    return ReactDOM.createPortal(
        <div className="RulesModal">
            <ModalContent />
        </div>
        , document.getElementById('portal') as HTMLElement)
}