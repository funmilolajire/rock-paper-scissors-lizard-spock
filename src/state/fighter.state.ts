import { createState, useState, State, DevTools } from '@hookstate/core';
import { outcome } from './utils/rules';
import { DevToolsInitialize } from '@hookstate/devtools';

export const fighters = ["rock", "paper", "scissors", "lizard", "spock"];

const playerFighter = createState('');
const houseFighter = createState('');

const wrapState = (state: State<string>) => ({
    get: () => state.value,
    select: (name: string) => state.set(prev => prev = name),
})

//to use outside of a component
export const accessPlayerFighter = () => wrapState(playerFighter);
export const accessHouseFighter = () => wrapState(houseFighter);

//to use within a component
export const usePlayerFighter = () => wrapState(useState(playerFighter));
export const useHouseFighter = () => wrapState(useState(houseFighter));

//devtools
DevTools(playerFighter).label('player-fighter')
DevTools(houseFighter).label('house-fighter')
DevToolsInitialize({
    monitored: ['player-fighter', 'house-fighter'],
    callstacksDepth: 30
})
