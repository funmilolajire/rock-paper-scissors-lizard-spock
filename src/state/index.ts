export * from './modal.state';
export * from './score.state';
export * from './fighter.state';