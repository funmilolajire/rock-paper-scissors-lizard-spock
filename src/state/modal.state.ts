import { createState, useState, State, DevTools } from '@hookstate/core';
import { DevToolsInitialize } from '@hookstate/devtools';

const modalState = createState(false);
const wrapState = (state: State<boolean>) => ({
    get: () => state.value,
    open: () => state.set(true),
    close: () => state.set(false)
})

//to use outside of a component
export const accessModalState = () => wrapState(modalState);

//to use within a component
export const useModalState = () => wrapState(useState(modalState));

//devtools
DevTools(modalState).label('modal-state')
DevToolsInitialize({
    monitored: ['modal-state'],
    callstacksDepth: 30
})
