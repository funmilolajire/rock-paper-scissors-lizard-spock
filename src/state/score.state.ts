import { createState, useState, State, DevTools } from '@hookstate/core';
import { DevToolsInitialize } from '@hookstate/devtools';
import { Persistence } from '@hookstate/persistence'


const playerState = createState(0);
const houseState = createState(0);

const wrapState = (state: State<number>) => ({
    get: () => state.value,
    increment: () => state.set(prev => prev + 1),
    clear: () => state.set(0)
})

//to use from outside a component
export const accessPlayerState = () => wrapState(playerState.attach(Persistence('playerScore')))
export const accessHouseState = () => wrapState(houseState.attach(Persistence('houseScore')))

//to use from within a component
export const usePlayerState = () => wrapState(useState(playerState).attach(Persistence('playerScore')))
export const useHouseState = () => wrapState(useState(houseState).attach(Persistence('houseScore')))

//devtools
DevTools(playerState).label('player-state')
DevTools(houseState).label('house-state')
DevToolsInitialize({
    monitored: ['player-state', 'house-state'],
    callstacksDepth: 30
})
