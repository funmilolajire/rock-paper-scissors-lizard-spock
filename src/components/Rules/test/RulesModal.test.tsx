import { mount } from 'enzyme';
import { RulesModal } from '../RulesModal';
import { accessModalState } from '../../../state/modal.state';

describe('rulesmodal component', () => {
    const portal = global.document.createElement('div')
    portal.setAttribute('id', 'portal')
    const body = global.document.querySelector('body')
    body?.appendChild(portal)
    let component = mount(<RulesModal />)
    test('make a snapshot of rulesmodal component', () => {
        expect(component).toMatchSnapshot()
    })
    test('modal closes on click', () => {
        const closeButton = component.find('.RulesModal .ModalContent .top span')
        closeButton.simulate('click')
        expect(accessModalState().get()).toBeFalse()
    })
})