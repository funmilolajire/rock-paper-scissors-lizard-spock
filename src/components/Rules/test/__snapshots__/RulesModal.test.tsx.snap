// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rulesmodal component make a snapshot of rulesmodal component 1`] = `
<RulesModal>
  <Portal
    containerInfo={
      <div
        id="portal"
      >
        <div
          class="RulesModal"
        >
          <div
            class="ModalContent"
          >
            <section
              class="top"
            >
              <h3>
                Rules
              </h3>
              <span
                class="close"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 512 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M405 136.798L375.202 107 256 226.202 136.798 107 107 136.798 226.202 256 107 375.202 136.798 405 256 285.798 375.202 405 405 375.202 285.798 256z"
                  />
                </svg>
              </span>
            </section>
            <figure>
              <img
                alt="rules"
                src="rules.svg"
              />
            </figure>
          </div>
        </div>
      </div>
    }
  >
    <div
      className="RulesModal"
    >
      <ModalContent>
        <div
          className="ModalContent"
        >
          <section
            className="top"
          >
            <h3>
              Rules
            </h3>
            <span
              className="close"
              onClick={[Function]}
            >
              <IoMdClose>
                <IconBase
                  attr={
                    Object {
                      "viewBox": "0 0 512 512",
                    }
                  }
                >
                  <svg
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    strokeWidth="0"
                    style={
                      Object {
                        "color": undefined,
                      }
                    }
                    viewBox="0 0 512 512"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M405 136.798L375.202 107 256 226.202 136.798 107 107 136.798 226.202 256 107 375.202 136.798 405 256 285.798 375.202 405 405 375.202 285.798 256z"
                      key="0"
                    />
                  </svg>
                </IconBase>
              </IoMdClose>
            </span>
          </section>
          <figure>
            <img
              alt="rules"
              src="rules.svg"
            />
          </figure>
        </div>
      </ModalContent>
    </div>
  </Portal>
</RulesModal>
`;
